### PR TITLE
Add CI for MSVC using dkml-workflows

### PR DESCRIPTION
**Pull request by jonahbeckford**
Implements #6535

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -22,6 +22,8 @@ Unreleased
 - Remove "Entering Directory" messages for `$ dune install`. (#6513,
   @rgrinberg)
 
+- Add CI testing with MSVC (todo, @jonahbeckford)
```
Comment (jonahbeckford) [inline CHANGES.md:25]:
```suggestion
- Add CI testing with MSVC (#6540, @jonahbeckford)
```

**Comment by rgrinberg** [comment]
Would it be possible to remove the stuff that doesn't pertain to msvc?

**Comment by jonahbeckford** [comment]
> Would it be possible to remove the stuff that doesn't pertain to msvc?

Is "the stuff" talking about the gh-darwin/gh-linux folders? Sure, we can remove files that were generated during dune promote with a Makefile rule. Is that fine?

**Comment by rgrinberg** [comment]
Yes, that is what I meant. Removing them in a makefile rule is fine

**Comment by rgrinberg** [comment]
@nojb can  you review please?

**Review by nojb — APPROVED** [review]
I have to confess that I find the amount of generated code a bit scary, but the results are good: the CI jobs (one for 32-bit and one for 64-bit) each take ~10m to complete once the cache is full.

The generated code itself is hard to review, but here are some general questions:

- What is the version of MSVC targeted?
- What is the workflow to update the generated files? `make update-dkml`?
- What is the required environment to execute the `desktop-ci-*` targets in the main `Makefile`?

For the record, could we enumerate somewhere which files are generated and which files are supposed to be maintained by hand? In particular, `build-test.sh` is hand-written, right?

I am approving based on the general idea that this is an strict improvement on the present situation (no MSVC testing at all).

**Comment by jonahbeckford** [comment]
> * What is the version of MSVC targeted?

There is no "target" per se. The CI machine (or your developer machine) is queried for all MSVC versions, and any versions that I know to be compatible with OCaml are used. The precise definition is `Get-CompatibleVisualStudios` at https://github.com/diskuv/dkml-runtime-distribution/blob/main/src/windows/Machine/Machine.psm1; that package is available for common use in the central Opam repository.

The compatible versions today are: Visual Studio 2019 with a) Windows 10 SDK (10.0.18362.0) and b) MSVC v142 - VS 2019 C++ x64/x86 build tools (v14.25) or (v14.26). *Machine.psm1 has notes why other MSVC 2019 versions are not compatible with OCaml; very much beyond the scope of this PR though!*

More importantly, just running it will show the user what is on their system and what Visual Studio components they are missing (if any).

> * What is the workflow to update the generated files? `make update-dkml`?

Yes. Obviously assuming there is an update available from `opam update`.

> * What is the required environment to execute the `desktop-ci-*` targets in the main `Makefile`?

Same as first answer.

> For the record, could we enumerate somewhere which files are generated and which files are supposed to be maintained by hand? In particular, `build-test.sh` is hand-written, right?

To make it easy I'll put a README.md in `ci/setup-dkml/`: only those files are auto-generated.

I'll add in that README a link to https://github.com/diskuv/dkml-workflows#readme (setup-dkml) which a) describes what is hand-generated and what goes into it, and b) includes auto-tested examples and c) implicitly defines a place for bug reports.

**Comment by nojb** [comment]
> I'll add in that README a link to https://github.com/diskuv/dkml-workflows#readme (setup-dkml) which a) describes what is hand-generated and what goes into it, and b) includes auto-tested examples and c) implicitly defines a place for bug reports.

Thanks, that's good enough for me! I sync'ed your PR with `main` and if the CI passes, let's merge.

**Pull request merged**